[cli][smoke-tests] Add command and test to update validator network addresses

ikabiljo opened the pull request:
extracted the same config from registering a new validator and used for the new command.
Seems like registering a new validator is broken on main, as bls proof of availability is missing,
but since I want to merge this to testnet branch (I assume?) that cannot be part of this diff.


### Description

### Test Plan
how should I test this properly?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/1946)
<!-- Reviewable:end -->

gregnazario commented on crates/aptos/src/node/mod.rs:
```diff
@@ -486,3 +495,52 @@ impl CliCommand<serde_json::Value> for ShowValidatorSet {
+                "update_network_and_fullnode_addresses",
+                vec![],
+                vec![
+                    bcs::to_bytes(&consensus_public_key)?,
```
I'm wondering, why do we need the consensus public key here?

zekun000 replied:
I don't think this should be the public_key, this is the pool address

ikabiljo replied:
yeah, misunderstood the param

gregnazario commented on crates/aptos/src/node/mod.rs:
```diff
@@ -288,6 +287,16 @@ impl RegisterValidatorCandidate {
     }
 }
 
+#[derive(Parser)]
+/// Register the current account as a Validator candidate
```
I'd flip these, keep the derive next to the function because the doc comments can be many lines.

zekun000 commented on crates/aptos/src/node/mod.rs:
```diff
@@ -288,6 +287,16 @@ impl RegisterValidatorCandidate {
+    pub(crate) txn_options: TransactionOptions,
+
+    #[clap(flatten)]
+    pub(crate) validator_config_args: ValidatorConfigArgs,
```
where does this config come from? how to update those values?

gregnazario replied:
This is in the PR.

I think since the consensus public key is not required, you shouldn't provide all of these args, but instead a subset

zekun000 replied:
hmm, but it reads from a file? how do we update that file?

ikabiljo replied:
look at lines 218-285 above. 
it allows people to set either:

  `Must provide either --validator-config-file or --validator-host`

gregnazario replied:
The file come straight out of the Genesis tooling.  You can update the file accordingly, as they're just simply named in yaml

gregnazario commented:
Where are we on this change?  Should I hijack this PR?

ikabiljo commented:
AIT2 didn't actually have the move code needed, so urgency was removed. 
Unless you need it urgently,  I'll get to it early next week.  I'm going to be adding back smoke tests for operational actions, including this one

ikabiljo commented:
I've updated the diff and added an unit test. 

Notes:
- Rest API results are not deserializable, so added manual deserialization via to_validator_config/to_validator_set
- faucet can make root LocalAccount get out of sync. Added function to re-sync it. It would be better to be changing sequence number from faucet, but that is a bit more convoluted (i.e. making sequence_number in local account be behind Arc, and pausing that through faucet), but not sure it is worth the complexity

movekevin commented on crates/aptos/src/node/mod.rs:
```diff
@@ -693,3 +695,61 @@ impl CliCommand<()> for RunLocalTestnet {
+                    bcs::to_bytes(&address)?,
+                    // Double BCS encode, so that we can hide the original type
+                    bcs::to_bytes(&bcs::to_bytes(&validator_network_addresses)?)?,
+                    bcs::to_bytes(&bcs::to_bytes(&full_node_network_addresses)?)?,
```
@gregnazario We had a fun dive last time regarding this. Is this a single address or a vector again?

gregnazario replied:
This is a double BCS encoded vector of addresses

movekevin commented on testsuite/smoke-test/src/aptos_cli.rs:
```diff
@@ -140,16 +163,100 @@ async fn test_register_and_update_validator() {
+    .await
+    .unwrap();
+
+    tokio::time::sleep(Duration::from_secs(2)).await;
```
Is this for waiting long enough so the tx is executed? 2 secs seems like a magic number

gregnazario replied:
We should probably work on a state based check vs just a raw sleep

ikabiljo replied:
this was unnecessary, left from previous testing. CliTestFramework wait for transactions results, and returns transactions, so should be good.

gregnazario commented on crates/aptos/src/node/mod.rs:
```diff
@@ -693,3 +695,61 @@ impl CliCommand<()> for RunLocalTestnet {
+            vec![validator_host.as_network_address(validator_network_public_key)?];
+        let full_node_network_addresses =
+            match (full_node_host.as_ref(), full_node_network_public_key) {
+                (Some(host), Some(public_key)) => vec![host.as_network_address(public_key)?],
+                _ => vec![],
```
Interesting, would we want to fail if it only has partial data?
